keps/369-job-interface/README.md

```diff
@@ -0,0 +1,423 @@
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+Part I: Job Interface
+
+We will define a new interface named GenericJob, this should be implemented by custom job-like applications:
+
+```golang
+type GenericJob interface {
+  // Object returns the job instance.
+  Object() client.Object
+  // IsSuspend returns whether the job is suspend or not.
+  IsSuspend() bool
+  // Suspend will suspend the job.
+  Suspend() error
+  // UnSuspend will unsuspend the job.
+  UnSuspend() error
+  // ResetStatus will reset the job status to the original state.
+  // If true, status is modified, if not, status is as it was.
+  ResetStatus() bool
+  // InjectNodeAffinity will inject the node affinity extracting from workload to job.
+  InjectNodeAffinity(nodeSelectors []map[string]string) error
+  // RestoreNodeAffinity will restore the original node affinity of job.
+  RestoreNodeAffinity(nodeSelectors []map[string]string) error
+  // Finished means whether the job is completed/failed or not,
+  // condition represents the workload finished condition.
+  Finished() (condition metav1.Condition, finished bool)
+  // PodSets will build workload podSets corresponding to the job.
+  PodSets() []kueue.PodSet
+  // EquivalentToWorkload validates whether the workload is semantically equal to the job.
+  EquivalentToWorkload(wl kueue.Workload) bool
+  // PriorityClass returns the job's priority class name.
+  PriorityClass() string
+  // QueueName returns the queue name the job enqueued.
+  QueueName() string
+  // Ignored instructs whether this job should be ignored in reconciling, e.g. lacking the queueName.
+  Ignored() bool
+}
+```
+
+Also defined another interface `SequenceJob` for scheduling in sequence requirements, see design details here [All-or-nothing semantics for job resource assignment](https://github.com/kubernetes-sigs/kueue/tree/main/keps/349-all-or-nothing#kep-349-all-or-nothing-semantics-for-job-resource-assignment).
+
+```golang
+type SequenceJob interface {
+    // PodsReady instructs whether job derived pods are all ready now.
+    PodsReady() bool
+}
+```
+
+Part II: Kubernetes Job implementation details
+
+We'll wrap the batchv1.Job to `BatchJob` who implements the GenericJob interface, as well as SequenceJob interface.
+
+```golang
+type BatchJob struct {
+  batchv1.Job
+}
+
+var _ GenericJob = &BatchJob{}
+var _ SequenceJob = &BatchJob{}
+```
+
+Part III: Library
+
+We'll have several helper functions in constructing controllers, they can be imported directly in the future:
```

Comment:
I think we can leave this to the code review. The pseudocode below should be enough